# 4.8: Grouping Data & Aggregating Variables

### Import previous exercise classified records from prepared data

In [1]:
import pandas as pd
import numpy as np
import os

pd.options.display.float_format = '{:.2f}'.format

class color:
   PURPLE = '\033[95m'
   CYAN = '\033[96m'
   DARKCYAN = '\033[36m'
   BLUE = '\033[94m'
   GREEN = '\033[92m'
   YELLOW = '\033[93m'
   RED = '\033[91m'
   BOLD = '\033[1m'
   UNDERLINE = '\033[4m'
   END = '\033[0m'
    
print(color.BLUE + 'Read orders_products_classified.pkl file' + color.END)

path = r'C:\\Users\\mutta\\Documents\\Project 4\\' 


ords_prods_merged_pkl = os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_classified.pkl')

df_ords_prods_merged  = pd.read_pickle(ords_prods_merged_pkl)

df_ords_prods_merged.head(10)

Read orders_products_classified.pkl file


order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   
5   2398795        1             2                   3                  7   
6   2398795        1             2                   3                  7   
7   2398795        1             2                   3                  7   
8   2398795        1             2                   3                  7   
9   2398795        1             2                   3                  7   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                    0.00         196                  1          0   
1                    0.00       14084                  2          0   
2                    0.00       12427                  3          0   
3                    0.00       26088                  4          0   
4                    0.00       26405                  5          0   
5                   15.00         196                  1          1   
6                   15.00       10258                  2          0   
7                   15.00       12427                  3          1   
8                   15.00       13176                  4          0   
9                   15.00       26088                  5          1   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda     77.00           7.00    9.00   
1  Organic Unsweetened Vanilla Almond Milk     91.00          16.00   12.50   
2                      Original Beef Jerky     23.00          19.00    4.40   
3               Aged White Cheddar Popcorn     23.00          19.00    4.70   
4         XL Pick-A-Size Paper Towel Rolls     54.00          17.00    1.00   
5                                     Soda     77.00           7.00    9.00   
6                               Pistachios    117.00          19.00    3.00   
7                      Original Beef Jerky     23.00          19.00    4.40   
8                   Bag of Organic Bananas     24.00           4.00   10.30   
9               Aged White Cheddar Popcorn     23.00          19.00    4.70   

         price_label     busiest_day busy_days_of_the_week  \
0  Mid-range product  Regularly busy          Regular days   
1  Mid-range product  Regularly busy          Regular days   
2  Low-range product  Regularly busy          Regular days   
3  Low-range product  Regularly busy          Regular days   
4  Low-range product  Regularly busy          Regular days   
5  Mid-range product  Regularly busy          slowest days   
6  Low-range product  Regularly busy          slowest days   
7  Low-range product  Regularly busy          slowest days   
8  Mid-range product  Regularly busy          slowest days   
9  Low-range product  Regularly busy          slowest days   

  busiest_period_of_day  
0        Average orders  
1        Average orders  
2        Average orders  
3        Average orders  
4        Average orders  
5        Average orders  
6        Average orders  
7        Average orders  
8        Average orders  
9        Average orders

In [2]:
df_ords_prods_merged.columns

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'product_id',
       'add_to_cart_order', 'reordered', 'product_name', 'aisle_id',
       'department_id', 'prices', 'price_label', 'busiest_day',
       'busy_days_of_the_week', 'busiest_period_of_day'],
      dtype='object')

### Aggregated mean on all columns for each department

In [3]:
print(color.BLUE + 'Aggregated mean of the numeric columns grouped by “department_id”' + color.END)

print('-------------------------------------------------------”')

group_value = df_ords_prods_merged.groupby('department_id').agg({'order_number': ['mean', 'min', 'max'],
                                                                'prices': ['mean', 'min', 'max'],
                                                                'orders_day_of_week': ['mean', 'min', 'max'],
                                                                'order_hour_of_day': ['mean', 'min', 'max'],
                                                                'days_since_prior_order': ['mean', 'min', 'max']
                                                                })
group_value

Aggregated mean of the numeric columns grouped by “department_id”
-------------------------------------------------------”


order_number         prices               orders_day_of_week  \
                      mean min max   mean  min      max               mean   
department_id                                                                
1.00                 15.46   1  99   7.74 1.00    15.00               2.79   
2.00                 17.28   1  99   6.99 1.10    15.00               2.85   
3.00                 17.17   1  99   7.85 1.00    15.00               2.74   
4.00                 17.81   1  99   7.98 1.00    15.00               2.68   
5.00                 15.22   1  99   8.14 1.00    15.00               3.16   
6.00                 16.44   1  99   7.68 1.00    15.00               2.69   
7.00                 17.23   1  99   7.68 1.00    15.00               2.81   
8.00                 15.34   1  99   7.89 1.00    15.00               2.93   
9.00                 15.90   1  99   7.35 1.00    15.00               2.69   
10.00                20.20   1  99   8.35 1.40    14.10               2.76   
11.00                16.17   1  99   8.00 1.00    15.00               2.82   
12.00                15.89   1  99  16.29 8.00    25.00               2.72   
13.00                16.58   1  99   8.01 1.00    20.00               2.81   
14.00                16.77   1  99   8.03 1.00    14.90               2.73   
15.00                16.17   1  99   7.54 1.00    15.00               2.71   
16.00                17.67   1  99  33.48 1.00 99999.00               2.74   
17.00                15.69   1  99   7.38 1.00    15.00               2.84   
18.00                19.31   1  99   7.64 1.00    15.00               2.71   
19.00                17.18   1  99   4.28 1.60     7.00               2.75   
20.00                16.47   1  99   7.78 1.00    15.00               2.71   
21.00                22.90   1  99   8.66 1.00    15.00               2.80   

                      order_hour_of_day         days_since_prior_order       \
              min max              mean min max                   mean  min   
department_id                                                                 
1.00            0   6             13.68   0  23                  10.90 0.00   
2.00            0   6             13.50   0  23                  10.24 0.00   
3.00            0   6             13.40   0  23                  10.39 0.00   
4.00            0   6             13.42   0  23                  10.19 0.00   
5.00            0   6             13.59   0  23                   9.49 0.00   
6.00            0   6             13.52   0  23                  10.50 0.00   
7.00            0   6             13.31   0  23                  10.36 0.00   
8.00            0   6             13.58   0  23                  11.06 0.00   
9.00            0   6             13.55   0  23                  10.95 0.00   
10.00           0   6             13.46   0  23                   9.59 0.00   
11.00           0   6             13.49   0  23                  10.78 0.00   
12.00           0   6             13.47   0  23                  10.67 0.00   
13.00           0   6             13.46   0  23                  10.40 0.00   
14.00           0   6             13.33   0  23                  10.54 0.00   
15.00           0   6             13.45   0  23                  10.85 0.00   
16.00           0   6             13.37   0  23                  10.31 0.00   
17.00           0   6             13.35   0  23                  11.15 0.00   
18.00           0   6             13.40   0  23                   9.47 0.00   
19.00           0   6             13.36   0  23                  10.26 0.00   
20.00           0   6             13.48   0  23                  10.59 0.00   
21.00           0   6             13.45   0  23                   9.88 0.00   

                     
                max  
department_id        
1.00          30.00  
2.00          30.00  
3.00          30.00  
4.00          30.00  
5.00          30.00  
6.00          30.00  
7.00          30.00  
8.00          30.00  
9.

### Analysis on grouped data

###### order_number	

                The department with an ID of 21(missing) & 10(bulk) has the average highest orders. 
                Conversely, the department with an ID of 5 (alcohol), has a 
                mean of around 15.22. The insight from this is that missing products name are 
                considerably more on average than any other goods.


###### prices	
                The department with an ID of 16 has the average highest sales 33.48.
                Also, the maximum sales value in department 16 has 99999.00.
                Conversely, the department with an ID of 19 has a least average sales value.

###### orders_day_of_week	
                All department made the highest sales on Sunday.


    

### Create Loyalty customers flag

“Loyalty” customers are those that are loyal to a brand. These are customers that come back time
and time again and use the service or buy the product often. You’ve been asked to locate these loyalty
customers at Instacart so that the business strategy team can employ some kind of bonus point program for them.

    
        If the maximum orders the user has made is over 40, then the customer will be labeled a “Loyal customer.”

        If the maximum orders the user has made is over 10 but less than or equal to 40, then the customer will 
        be labeled a “Regular customer.”

        If the maximum orders the user has made is less than or equal to 10, then the customer will be 
        labeled a “New customer.”

In [4]:
df_ords_prods_merged['max_order'] = df_ords_prods_merged.groupby(['user_id'])['order_number'].transform(np.max)


df_ords_prods_merged.loc[df_ords_prods_merged['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'


df_ords_prods_merged.loc[(df_ords_prods_merged['max_order'] <= 40) & 
                         (df_ords_prods_merged['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'


df_ords_prods_merged.loc[df_ords_prods_merged['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [5]:
stat_1 = df_ords_prods_merged.groupby(['loyalty_flag']).agg({'prices': ['mean', 'min', 'max','sum']})


In [6]:
print(color.BLUE + 'The total Loyal categories counts' + color.END)

print(df_ords_prods_merged['loyalty_flag'].value_counts())

print('-------------------------------------------------------------------------------')

print(color.BLUE + 'The basic statistics of the product prices for each customer categories' + color.END)

print(stat_1 )

print('-------------------------------------------------------------------------------')


The total Loyal categories counts
Regular customer    15891507
Loyal customer      10294027
New customer         6249525
Name: loyalty_flag, dtype: int64
-------------------------------------------------------------------------------
The basic statistics of the product prices for each customer categories
                 prices                           
                   mean  min      max          sum
loyalty_flag                                      
Loyal customer    10.39 1.00 99999.00 106814042.20
New customer      13.29 1.00 99999.00  83011787.20
Regular customer  12.50 1.00 99999.00 198391693.20
-------------------------------------------------------------------------------


### Analysis on Loyal Customer


        The average purchase order of new customer is more compared to the regular and royal customers.
        
        
        Total purchase sale value of regular customer is 198,391,693.20 which much more than the other categories.


### Create a spending flag for each user

The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:

        If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
        
        If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [7]:
df_ords_prods_merged['max_product_orders'] = df_ords_prods_merged.groupby(['user_id'])['prices'].transform(np.mean)


df_ords_prods_merged.loc[df_ords_prods_merged['max_product_orders'] < 10, 'spending_flag'] = 'Low spender'

df_ords_prods_merged.loc[df_ords_prods_merged['max_product_orders'] >= 10, 'spending_flag'] = 'High spender'

In [8]:
df_ords_prods_merged.loc[df_ords_prods_merged['user_id'] == 2]

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
59    2168274        2             1                   2                 11   
60    2168274        2             1                   2                 11   
61    2168274        2             1                   2                 11   
62    2168274        2             1                   2                 11   
63    2168274        2             1                   2                 11   
..        ...      ...           ...                 ...                ...   
249    839880        2            14                   3                 10   
250    839880        2            14                   3                 10   
251    839880        2            14                   3                 10   
252    839880        2            14                   3                 10   
253    839880        2            14                   3                 10   

     days_since_prior_order  product_id  add_to_cart_order  reordered  \
59                     0.00       32792                  1          0   
60                     0.00       47766                  2          0   
61                     0.00       20574                  3          0   
62                     0.00       12000                  4          0   
63                     0.00       48110                  5          0   
..                      ...         ...                ...        ...   
249                   13.00        2573                 12          1   
250                   13.00       39928                 13          0   
251                   13.00       20785                 14          0   
252                   13.00       24768                 15          0   
253                   13.00        7963                 16          0   

                                 product_name  ...  department_id  prices  \
59             Chipotle Beef & Pork Realstick  ...          19.00    5.20   
60                            Organic Avocado  ...           4.00    6.30   
61                             Roasted Turkey  ...          20.00    2.30   
62   Baked Organic Sea Salt Crunchy Pea Snack  ...          13.00   10.90   
63    Thin Stackers Brown Rice Lightly Salted  ...          19.00    2.70   
..                                        ...  ...            ...     ...   
249                       Garlic Pepper Ramen  ...           6.00   13.70   
250                              Organic Kiwi  ...           4.00    1.70   
251    Extra Sharp Aged Cheddar Cheese slices  ...          16.00   10.10   
252           All Natural Honey Almond Butter  ...          13.00    6.50   
253             Gluten Free Whole Grain Bread  ...           1.00    5.30   

           price_label     busiest_day busy_days_of_the_week  \
59   Mid-range product  Regularly busy          Regular days   
60   Mid-range product  Regularly busy          Regular days   
61   Low-range product  Regularly busy          Regular days   
62   Mid-range product  Regularly busy          Regular days   
63   Low-range product  Regularly busy          Regular days   
..                 ...             ...                   ...   
249  Mid-range product  Regularly busy          slowest days   
250  Low-range product  Regularly busy          slowest days   
251  Mid-range product  Regularly busy          slowest days   
252  Mid-range product  Regularly busy          slowest days   
253  Mid-range product  Regularly busy          slowest days   

    busiest_period_of_day max_order      loyalty_flag max_product_orders  \
59            Most orders        14  Regular customer               7.52   
60            Most orders        14  Regular customer               7.52   
61            Most orders        14  Regular customer               7.52   
62            Most orders        14  Regular customer               7.52   
63            Most orders        14  Regular customer               7.52   
..                    ...    

In [9]:
spending_stat = df_ords_prods_merged.groupby(['spending_flag']).agg({'prices': ['mean', 'min', 'max','sum']})



In [10]:
print(color.BLUE + 'The total Spending categories counts' + color.END)

print(df_ords_prods_merged['spending_flag'].value_counts())

print('-------------------------------------------------------------------------------')

print(color.BLUE + 'The basic statistics of the spending categories' + color.END)

print(spending_stat)

print('-------------------------------------------------------------------------------')


The total Spending categories counts
Low spender     31800381
High spender      634678
Name: spending_flag, dtype: int64
-------------------------------------------------------------------------------
The basic statistics of the spending categories
              prices                           
                mean  min      max          sum
spending_flag                                  
High spender  222.34 1.00 99999.00 141020378.80
Low spender     7.78 1.00    25.00 247197143.80
-------------------------------------------------------------------------------


### Create an order frequency flag that marks the regularity of a user

In order to send relevant notifications to users within the app (for instance, asking users
if they want to buy the same item again), the Instacart team wants you to determine frequent
versus non-frequent customers. Create an order frequency flag that marks the regularity of a 
user’s ordering behavior according to the median in the “days_since_prior_order” column. 
The criteria for the flag should be as follows:
    
        If the median of “days_since_prior_order” is higher than 20, then the customer should be
        labeled a “Non-frequent customer.”
        
        If the median is higher than 10 and lower than or equal to 20, then the customer should 
        be labeled a “Regular customer.”
        
        If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [11]:
df_ords_prods_merged['median_of_frequency'] = df_ords_prods_merged.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


df_ords_prods_merged.loc[df_ords_prods_merged['median_of_frequency'] > 20, 'frequency_flag'] = 'Non-frequent customer'


df_ords_prods_merged.loc[(df_ords_prods_merged['median_of_frequency'] <= 20) & 
                         (df_ords_prods_merged['median_of_frequency'] > 10), 'frequency_flag'] = 'Regular customer'


df_ords_prods_merged.loc[df_ords_prods_merged['median_of_frequency'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [12]:
frequency_stat = df_ords_prods_merged.groupby(['frequency_flag']).agg({'days_since_prior_order': ['mean', 'median', 'min', 'max', 'count']})


In [13]:
print(color.BLUE + 'The total frequency categories counts' + color.END)

print(df_ords_prods_merged['frequency_flag'].value_counts())

print('-------------------------------------------------------------------------------')

print(color.BLUE + 'The basic statistics of days since prior order for frequent customer categories' + color.END)

print(frequency_stat)

print('-------------------------------------------------------------------------------')


The total frequency categories counts
Frequent customer        22816041
Regular customer          6929012
Non-frequent customer     2690006
Name: frequency_flag, dtype: int64
-------------------------------------------------------------------------------
The basic statistics of days since prior order for frequent customer categories
                      days_since_prior_order                            
                                        mean median  min   max     count
frequency_flag                                                          
Frequent customer                       7.80   6.00 0.00 30.00  22816041
Non-frequent customer                  20.71  27.00 0.00 30.00   2690006
Regular customer                       14.91  14.00 0.00 30.00   6929012
-------------------------------------------------------------------------------


In [14]:
df_ords_prods_merged.head(10)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   
5   2398795        1             2                   3                  7   
6   2398795        1             2                   3                  7   
7   2398795        1             2                   3                  7   
8   2398795        1             2                   3                  7   
9   2398795        1             2                   3                  7   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                    0.00         196                  1          0   
1                    0.00       14084                  2          0   
2                    0.00       12427                  3          0   
3                    0.00       26088                  4          0   
4                    0.00       26405                  5          0   
5                   15.00         196                  1          1   
6                   15.00       10258                  2          0   
7                   15.00       12427                  3          1   
8                   15.00       13176                  4          0   
9                   15.00       26088                  5          1   

                              product_name  ...        price_label  \
0                                     Soda  ...  Mid-range product   
1  Organic Unsweetened Vanilla Almond Milk  ...  Mid-range product   
2                      Original Beef Jerky  ...  Low-range product   
3               Aged White Cheddar Popcorn  ...  Low-range product   
4         XL Pick-A-Size Paper Towel Rolls  ...  Low-range product   
5                                     Soda  ...  Mid-range product   
6                               Pistachios  ...  Low-range product   
7                      Original Beef Jerky  ...  Low-range product   
8                   Bag of Organic Bananas  ...  Mid-range product   
9               Aged White Cheddar Popcorn  ...  Low-range product   

      busiest_day  busy_days_of_the_week busiest_period_of_day max_order  \
0  Regularly busy           Regular days        Average orders        10   
1  Regularly busy           Regular days        Average orders        10   
2  Regularly busy           Regular days        Average orders        10   
3  Regularly busy           Regular days        Average orders        10   
4  Regularly busy           Regular days        Average orders        10   
5  Regularly busy           slowest days        Average orders        10   
6  Regularly busy           slowest days        Average orders        10   
7  Regularly busy           slowest days        Average orders        10   
8  Regularly busy           slowest days        Average orders        10   
9  Regularly busy           slowest days        Average orders        10   

   loyalty_flag max_product_orders  spending_flag median_of_frequency  \
0  New customer               6.37    Low spender               20.00   
1  New customer               6.37    Low spender               20.00   
2  New customer               6.37    Low spender               20.00   
3  New customer               6.37    Low spender               20.00   
4  New customer               6.37    Low spender               20.00   
5  New customer               6.37    Low spender               20.00   
6  New customer               6.37    Low spender               20.00   
7  New customer               6.37    Low spender               20.00   
8  New customer               6.37    Low spender               20.00   
9  New customer               6.37 

### Print summary of dataframe and export dataframe to pkl fie

In [15]:
print(color.BLUE + color.BOLD + 'Columns in dataframe' + color.END)
print(df_ords_prods_merged.columns)

print('------------------------------------------')
print(color.BLUE + color.BOLD + '\nNumber of rows in dataframe:' + color.END, df_ords_prods_merged.shape[0])

print(color.BLUE + color.BOLD + '\nNumber of columns in dataframe:' + color.END, df_ords_prods_merged.shape[1])

print('------------------------------------------')
df_ords_prods_merged.head(10)

Columns in dataframe
Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'product_id',
       'add_to_cart_order', 'reordered', 'product_name', 'aisle_id',
       'department_id', 'prices', 'price_label', 'busiest_day',
       'busy_days_of_the_week', 'busiest_period_of_day', 'max_order',
       'loyalty_flag', 'max_product_orders', 'spending_flag',
       'median_of_frequency', 'frequency_flag'],
      dtype='object')
------------------------------------------

Number of rows in dataframe: 32435059

Number of columns in dataframe: 23
------------------------------------------


order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   
5   2398795        1             2                   3                  7   
6   2398795        1             2                   3                  7   
7   2398795        1             2                   3                  7   
8   2398795        1             2                   3                  7   
9   2398795        1             2                   3                  7   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                    0.00         196                  1          0   
1                    0.00       14084                  2          0   
2                    0.00       12427                  3          0   
3                    0.00       26088                  4          0   
4                    0.00       26405                  5          0   
5                   15.00         196                  1          1   
6                   15.00       10258                  2          0   
7                   15.00       12427                  3          1   
8                   15.00       13176                  4          0   
9                   15.00       26088                  5          1   

                              product_name  ...        price_label  \
0                                     Soda  ...  Mid-range product   
1  Organic Unsweetened Vanilla Almond Milk  ...  Mid-range product   
2                      Original Beef Jerky  ...  Low-range product   
3               Aged White Cheddar Popcorn  ...  Low-range product   
4         XL Pick-A-Size Paper Towel Rolls  ...  Low-range product   
5                                     Soda  ...  Mid-range product   
6                               Pistachios  ...  Low-range product   
7                      Original Beef Jerky  ...  Low-range product   
8                   Bag of Organic Bananas  ...  Mid-range product   
9               Aged White Cheddar Popcorn  ...  Low-range product   

      busiest_day  busy_days_of_the_week busiest_period_of_day max_order  \
0  Regularly busy           Regular days        Average orders        10   
1  Regularly busy           Regular days        Average orders        10   
2  Regularly busy           Regular days        Average orders        10   
3  Regularly busy           Regular days        Average orders        10   
4  Regularly busy           Regular days        Average orders        10   
5  Regularly busy           slowest days        Average orders        10   
6  Regularly busy           slowest days        Average orders        10   
7  Regularly busy           slowest days        Average orders        10   
8  Regularly busy           slowest days        Average orders        10   
9  Regularly busy           slowest days        Average orders        10   

   loyalty_flag max_product_orders  spending_flag median_of_frequency  \
0  New customer               6.37    Low spender               20.00   
1  New customer               6.37    Low spender               20.00   
2  New customer               6.37    Low spender               20.00   
3  New customer               6.37    Low spender               20.00   
4  New customer               6.37    Low spender               20.00   
5  New customer               6.37    Low spender               20.00   
6  New customer               6.37    Low spender               20.00   
7  New customer               6.37    Low spender               20.00   
8  New customer               6.37    Low spender               20.00   
9  New customer               6.37 

In [16]:
# Export data to pkl

print('Saving the dataframe..')

df_ords_prods_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_grouped.pkl'))

print(color.BLUE + 'Data exported successfully' + color.END)




Saving the dataframe..
Data exported successfully
